### PR TITLE
Add support for Redis Sentinel

### PR DESCRIPTION
**Pull request by bekriebel**
This is a rebase of #232 from @heilerich onto the current mainline. I don't have a sentinel install available at the moment to test myself, but my client will be testing this out later in the week.

add client sentinel configuration and client init
rewire
add configuration sample

### CODE REVIEW

File: pkg/service/wire.go [hunk 1]
```diff
@@ -109,25 +109,54 @@ func createRedisClient(conf *config.Config) (*redis.Client, error) {
...
-		Username: conf.Redis.Username,
-		Password: conf.Redis.Password,
-		DB:       conf.Redis.DB,
-	}
-	if conf.Redis.UseTLS {
-		rcOptions = &redis.Options{
+	var rc *redis.Client
+	if conf.UseSentinel() {
+		logger.Infow("using multi-node routing via redis", "sentinel", true, "addr", conf.Redis.SentinelAddresses, "masterName", conf.Redis.MasterName)
+		rcOptions := &redis.FailoverOptions{
```
Comment (davidzhao) [inline pkg/service/wire.go:115]: can we reduce the amount of replication here? it seems that the only difference is TLS?

**Review by davidzhao — COMMENTED** [review]
Thanks for porting this over. I'm supportive of the change, but also have no way of validating if it works.

**Comment by bekriebel** [comment]
I'll comment back after my client gets a chance to try it out this week.

**Comment by kakaroto** [comment]
I am said client, and I tested the rebased build today, worked great, and I was able to randomly kill the redis master/replicas without any issues, livekit would just show debug messages saying connection lost to master and that a new master was elected.
See test performed below (using https://livekit.io/connection-test) in the middle of which I killed the master redis node, and there was no issues with the test, and the "Resuming connection after interruption" was successful too : 
image

This is a much better behavior than my current master-replicas setup which constantly fails due to kubernetes moving pods around during scaling operations. My livekit server has had 109 restarts in the last 6 days because of that, according to the k8s pod status. This PR will fix that, so thank you for considering it.

Reviewing the changes, I see a typo in the sample config yaml (`sentinel_addresses` instead of `sentinel_password`) but the most problematic issue for me was having to put the password in the config instead of being able to override it with environment variables like with `REDIS_PASSWORD`. I've told @bekriebel about the issue and it should be an easy fix to add the cli argument/env in `server/main.go` for the sentinel arguments.
I'll wait for the new build to test it with the env var, and confirm if I notice anything else from my side 👍

**Comment by bekriebel** [comment]
Thanks for responding here, @kakaroto! I'll get the typo fix and env variable changes in on Friday and push a new version for review then.

File: pkg/service/wire.go (hunk 1 above)
Comment (bekriebel) [reply pkg/service/wire.go:123]: Updated in the newer push.

**Comment by bekriebel** [comment]
All requested changes have been made. It's running on my test server with a single-node redis instance, and I believe @kakaroto will be testing it with sentinel soon

**Comment by kakaroto** [comment]
Reviewed the code and all looked good to me (though it's best to have someone familiar with the code review it again of course).
Tested the new build and it worked beautifully. Glad I can set the password in a `Secret` now and pass it through an env var. Also great to be able to override the `LIVEKIT_TURN_DOMAIN` via env var, thanks for adding that as well 👍

**Comment by davidzhao** [comment]
I'm not convinced that adding so many CLI args is a good idea. We recommend setting config via the LIVEKIT_CONFIG env variable, with YAML format. Is there a reason why that approach doesn't work?

**Comment by kakaroto** [comment]
> I'm not convinced that adding so many CLI args is a good idea. We recommend setting config via the LIVEKIT_CONFIG env variable, with YAML format. Is there a reason why that approach doesn't work?

The only one that's needed is the password and the turn domain (in my case). I think @bekriebel added the others only for completeness sake, likely thinking it was an oversight.

The issue with having a `LIVEKIT_CONFIG` with everything in it is that you need a single string with all the information. In a multi-region/multi-domain setup on kubernetes for example, it would mean having a duplicate `ConfigMap`  with all of the settings (which risks getting out of sync if we're not careful whenever we change something), whereas I prefer to have the basic config in the `LIVEKIT_CONFIG` and then set the password via a separate env var (stored in a `Secret` instead) and override the `LIVEKIT_TURN_DOMAIN` with a separate config. (And I just realized that one of my deployment had the TURN domain set to the domain from another deployment in a different region which is bad).
I don't think having them as CLI options is a problem as long as it's only to override the options already set in yaml. Most people would likely use the yaml, but having the option to override them with CLI or env vars can only be good in my opinion.

**Comment by davidzhao** [comment]
> The only one that's needed is the password and the turn domain (in my case). I think @bekriebel added the others only for completeness sake, likely thinking it was an oversight.

It wasn't an oversight. CLI is an interface to the server and each new argument carries a bit of cognitive load for the new user. (see `livekit-server --help`)

we already have too many options (IMHO) and I'm actively thinking of ways of reducing them. 

> The issue with having a `LIVEKIT_CONFIG` with everything in it is that you need a single string with all the information. In a multi-region/multi-domain setup on kubernetes for example, it would mean having a duplicate `ConfigMap` with all of the settings (which risks getting out of sync if we're not careful whenever we change something

I understand the desire to avoid duplication, but I'm not convinced it outweighs the burden to a novice user. Let's split this PR:  I'm in favor of merging sentinel support immediately. Adding args needs additional consideration.

**Comment by kakaroto** [comment]
Alright, I can understand that!
Can we get the redis_sentinel_password to default to the `redis_password` in that case? I am actually not sure why the username/password need to be different arguments, since you either use sentinel or you're not, it's still the same redis server/user/pass, it just uses failover options.
I'd be in favor of dropping those two parameters and making the sentinel options use `redis-username` and `redis-password` directly instead of `redis-sentinel-username` and `redis-sentinel-password`. That would simplify things as well.

I would then vote for a separate PR which adds the `TURN_DOMAIN` as the sole additional CLI parameter for overriding the config.
Thanks!

**Comment by bekriebel** [comment]
Happy to split this out and will do so. I do agree that the way config is current managed is a bit unwieldy and could use a rethinking.

I have a couple of proposals, but we can discuss those in the split PR.

1. At minimum, include the password fields and TURN Domain in settings. This would let key values that should be different per-environment be easily set on the CLI.
2. Put ALL settings into the CLI, but put `Hidden: True` on all but the most common ones. This would reduce the cognitive load of the `--help` output, but would still leave the option open for setting the values for those who need to.
3. Switching to a more full-fledged config option like Viper or using urfave cli's values from file options to make everything consistent.

I probably don't have time to do the third option right now, but could put the effort into the first or second option.

**Comment by bekriebel** [comment]
> Can we get the `redis_sentinel_password` to default to the `redis_password` in that case? I am actually not sure why the username/password need to be different arguments, since you either use sentinel or you're not, it's still the same redis server/user/pass, it just uses failover options.
> I'd be in favor of dropping those two parameters and making the sentinel options use `redis-username` and `redis-password` directly instead of `redis-sentinel-username` and `redis-sentinel-password`. That would simplify things as well.

I like this option and will go ahead and implement it this way if there's no objection from @davidzhao. It was done this way in the original PR, but I agree that there's no reason these can't be shared values since it is either/or.

**Comment by bekriebel** [comment]
> > I'd be in favor of dropping those two parameters and making the sentinel options use `redis-username` and `redis-password` directly instead of `redis-sentinel-username` and `redis-sentinel-password`. That would simplify things as well.
> 
> I like this option and will go ahead and implement it this way if there's no objection from @davidzhao. It was done this way in the original PR, but I agree that there's no reason these can't be shared values since it is either/or.

After testing, it turns out this shouldn't be done. The two username/password values are used for different settings and may not be the same in everyone's environment.

**Comment by bekriebel** [comment]
@davidzhao I've split this out as requested and opened up a new PR (#731) that adds the CLI values in with the `Hidden` option to avoid the help page clutter. This PR should be ready to go, but we can carry on the broader discussion about the options in #730.

**Review by davidzhao — APPROVED** [review]
Looks good @bekriebel! Thanks for following through on this one!

**Pull request merged**